In [ ]:
import pandas as pd
import json
import os
import numpy as np
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint
from sklearn.preprocessing import MinMaxScaler
from sklearn import tree

import seaborn as sns

import matplotlib.pyplot as plt
%matplotlib inline
pd.set_option('display.max_columns', 500)

#Modeling on Champion Names of each Match

In [ ]:
mcn = pd.read_csv('match_champ_names.csv')
mcn

Match ID  Team A Won?     A Top  A Jungle        A Mid  \
0       4848223288            1       Jax      Nunu       Syndra   
1       4848131748            1      Yone     Poppy  TwistedFate   
2       4848087528            1       Jax     Viego      Orianna   
3       4847265470            0      Azir      Kayn     Vladimir   
4       4848043757            1    Rumble   Kindred          Zoe   
...            ...          ...       ...       ...          ...   
128040  4814345455            1    Aatrox  Nocturne        Sylas   
128041  4813002074            0  Renekton   Warwick        Ziggs   
128042  4819650090            0       Jax    Graves        Nasus   
128043  4812981843            1     Garen       Zac         Azir   
128044  4814686156            1  Malphite   Trundle        Ziggs   

               A Bot A Support       B Top    B Jungle     B Mid     B Bot  \
0              Vayne     Shaco       Akali         Zac    Qiyana    Twitch   
1             Draven     Milio     Chogath  MonkeyKing     Sylas   Kalista   
2            Kalista      Sona        Shen        Udyr     Neeko      Jinx   
3               Jhin   Alistar       Garen    Nocturne       Zed     Kaisa   
4            Caitlyn     Senna  Tryndamere      Khazix  Vladimir     Sivir   
...              ...       ...         ...         ...       ...       ...   
128040  Heimerdinger     Senna       Garen      Maokai    Veigar    Twitch   
128041        Twitch   Alistar      KSante       Briar    Xerath  Tristana   
128042  Heimerdinger      Bard     Chogath     Trundle    Irelia     Vayne   
128043          Ashe      Lulu        Kled      Graves      Ekko  Tristana   
128044   MissFortune    Twitch  Tryndamere     Nidalee     Annie  Tristana   

        B Support  
0           Rakan  
1          Renata  
2          Thresh  
3           Rakan  
4         Alistar  
...           ...  
128040     Thresh  
128041  Seraphine  
128042      Milio  
128043       Nami  
128044   Nautilus  

[128045 rows x 12 columns]

In [ ]:
dummied = pd.get_dummies(mcn)
dummied

Match ID  Team A Won?  A Top_Aatrox  A Top_Ahri  A Top_Akali  \
0       4848223288            1             0           0            0   
1       4848131748            1             0           0            0   
2       4848087528            1             0           0            0   
3       4847265470            0             0           0            0   
4       4848043757            1             0           0            0   
...            ...          ...           ...         ...          ...   
128040  4814345455            1             1           0            0   
128041  4813002074            0             0           0            0   
128042  4819650090            0             0           0            0   
128043  4812981843            1             0           0            0   
128044  4814686156            1             0           0            0   

        A Top_Akshan  A Top_Alistar  A Top_Amumu  A Top_Anivia  A Top_Annie  \
0                  0              0            0             0            0   
1                  0              0            0             0            0   
2                  0              0            0             0            0   
3                  0              0            0             0            0   
4                  0              0            0             0            0   
...              ...            ...          ...           ...          ...   
128040             0              0            0             0            0   
128041             0              0            0             0            0   
128042             0              0            0             0            0   
128043             0              0            0             0            0   
128044             0              0            0             0            0   

        A Top_Aphelios  A Top_Ashe  A Top_AurelionSol  A Top_Azir  A Top_Bard  \
0                    0           0                  0           0           0   
1                    0           0                  0           0           0   
2                    0           0                  0           0           0   
3                    0           0                  0           1           0   
4                    0           0                  0           0           0   
...                ...         ...                ...         ...         ...   
128040               0           0                  0           0           0   
128041               0           0                  0           0           0   
128042               0           0                  0           0           0   
128043               0           0                  0           0           0   
128044               0           0                  0           0           0   

        A Top_Belveth  A Top_Blitzcrank  A Top_Brand  A Top_Braum  \
0                   0                 0            0            0   
1                   0                 0            0            0   
2                   0                 0            0            0   
3                   0                 0            0            0   
4                   0                 0            0            0   
...               ...               ...          ...          ...   
128040              0                 0            0            0   
128041              0                 0            0            0   
128042              0                 0            0            0   
128043              0                 0            0            0   
128044              0                 0            0            0   

        A Top_Briar  A Top_Caitlyn  A Top_Camille  A Top_Cassiopeia  \
0                 0              0              0                 0   
1                 0              0              0                 0   
2                 0              0              0                 0   
3                 0              0              0                 0   
4              

In [ ]:
X = dummied.drop(['Match ID', 'Team A Won?'], axis=1)
y = dummied['Team A Won?']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=42)

X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size = 0.2, random_state=42)

In [ ]:
mlp = MLPClassifier(random_state=42).fit(X_train, y_train)
mlp.score(X_val, y_val)

0.5106733772995488

In [ ]:
clf = tree.DecisionTreeClassifier(random_state=42).fit(X_train, y_train)
clf.score(X_val, y_val)

0.5080604674513304

In [ ]:
clf = RandomForestClassifier(max_depth = None, random_state=42).fit(X_train, y_train)
clf.score(X_val, y_val)

0.516877820201319

1. Modeling on Champion Names of each Match

  decison tree - 0.5080604674513304

  random forest - 0.516877820201319

#Modeling on Champion Win Rate in their Respective Role

In [ ]:
champ_wr = pd.read_csv('match_champ_winrate.csv')
champ_wr

Match ID  Team A Won?     A Top  A Jungle     A Mid     A Bot  \
0       4848223288            1  0.521313  0.503173  0.503341  0.516784   
1       4848131748            1  0.479153  0.483957  0.505706  0.519829   
2       4848087528            1  0.521313  0.485958  0.493604  0.510774   
3       4847265470            0  0.549865  0.503162  0.506785  0.494525   
4       4848043757            1  0.493161  0.528064  0.514953  0.483096   
...            ...          ...       ...       ...       ...       ...   
128040  4814345455            1  0.487486  0.486515  0.520461  0.472028   
128041  4813002074            0  0.489415  0.470032  0.485156  0.522730   
128042  4819650090            0  0.521313  0.517958  0.503480  0.472028   
128043  4812981843            1  0.490364  0.505829  0.488907  0.527675   
128044  4814686156            1  0.501658  0.484444  0.485156  0.503527   

        A Support     B Top  B Jungle     B Mid     B Bot  B Support  
0        0.472906  0.497863  0.505829  0.502770  0.522730   0.506754  
1        0.494923  0.502588  0.466134  0.520461  0.510774   0.506617  
2        0.512271  0.497370  0.508568  0.506992  0.507348   0.514714  
3        0.509598  0.490364  0.486515  0.503803  0.480917   0.506754  
4        0.517835  0.523136  0.495181  0.506785  0.482210   0.509598  
...           ...       ...       ...       ...       ...        ...  
128040   0.517835  0.490364  0.477379  0.464943  0.522730   0.514714  
128041   0.509598  0.510396  0.489342  0.471320  0.521180   0.478411  
128042   0.521975  0.502588  0.484444  0.496939  0.516784   0.494923  
128043   0.483742  0.518345  0.517958  0.505618  0.521180   0.483875  
128044   0.492683  0.523136  0.523916  0.500270  0.521180   0.491969  

[128045 rows x 12 columns]

In [ ]:
X = champ_wr.drop(['Match ID', 'Team A Won?'], axis=1)
y = champ_wr['Team A Won?']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=42)

X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size = 0.2, random_state=42)

In [ ]:
mlp = MLPClassifier(random_state=42).fit(X_train, y_train)
mlp.score(X_val, y_val)

0.5485580409438278

In [ ]:
clf = tree.DecisionTreeClassifier(random_state=42).fit(X_train, y_train)
clf.score(X_val, y_val)

0.5059965415295364

In [ ]:
clf = RandomForestClassifier(max_depth = None, random_state=42).fit(X_train, y_train)
clf.score(X_val, y_val)

0.5286997266692698

In [ ]:
knn = KNeighborsClassifier().fit(X_train, y_train)
knn.score(X_val, y_val)

0.5158699168851453

2. Modeling on Champion Win Rate in their Respective Role

  nn - 0.5485580409438278

  rf - 0.5286997266692698

  knn - 0.5158699168851453

#Modeling on Champion and Bot/Support Duo Win Rates

In [ ]:
champ_duo_wr = pd.read_csv('match_champ_duo_winrate.csv')
champ_duo_wr

Match ID  Team A Won?     A Top  A Jungle     A Mid     A Bot  \
0       4848223288            1  0.521313  0.503173  0.503341  0.516784   
1       4848131748            1  0.479153  0.483957  0.505706  0.519829   
2       4848087528            1  0.521313  0.485958  0.493604  0.510774   
3       4847265470            0  0.549865  0.503162  0.506785  0.494525   
4       4848043757            1  0.493161  0.528064  0.514953  0.483096   
...            ...          ...       ...       ...       ...       ...   
128040  4814345455            1  0.487486  0.486515  0.520461  0.472028   
128041  4813002074            0  0.489415  0.470032  0.485156  0.522730   
128042  4819650090            0  0.521313  0.517958  0.503480  0.472028   
128043  4812981843            1  0.490364  0.505829  0.488907  0.527675   
128044  4814686156            1  0.501658  0.484444  0.485156  0.503527   

        A Support     B Top  B Jungle     B Mid     B Bot  B Support  \
0        0.472906  0.497863  0.505829  0.502770  0.522730   0.506754   
1        0.494923  0.502588  0.466134  0.520461  0.510774   0.506617   
2        0.512271  0.497370  0.508568  0.506992  0.507348   0.514714   
3        0.509598  0.490364  0.486515  0.503803  0.480917   0.506754   
4        0.517835  0.523136  0.495181  0.506785  0.482210   0.509598   
...           ...       ...       ...       ...       ...        ...   
128040   0.517835  0.490364  0.477379  0.464943  0.522730   0.514714   
128041   0.509598  0.510396  0.489342  0.471320  0.521180   0.478411   
128042   0.521975  0.502588  0.484444  0.496939  0.516784   0.494923   
128043   0.483742  0.518345  0.517958  0.505618  0.521180   0.483875   
128044   0.492683  0.523136  0.523916  0.500270  0.521180   0.491969   

           A Duo     B Duo  
0       0.551724  0.554307  
1       0.536680  0.498225  
2       0.596154  0.550744  
3       0.510638  0.468321  
4       0.472785  0.524064  
...          ...       ...  
128040  0.433333  0.526749  
128041  0.493007  0.538462  
128042  0.250000  0.521595  
128043  0.495050  0.494253  
128044  0.428571  0.491379  

[128045 rows x 14 columns]

In [ ]:
X = champ_duo_wr.drop(['Match ID', 'Team A Won?'], axis=1)
y = champ_duo_wr['Team A Won?']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=42)

X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size = 0.2, random_state=42)

In [ ]:
mlp = MLPClassifier(random_state=42).fit(X_train, y_train)
mlp.score(X_val, y_val)

0.580632565404139

In [ ]:
clf = RandomForestClassifier(max_depth = 5, random_state=42).fit(X_train, y_train)
clf.score(X_val, y_val)

0.5734367155686952

In [ ]:
knn = KNeighborsClassifier().fit(X_train, y_train)
knn.score(X_val, y_val)

0.536620739666425

3. Modeling on Champion and Bot/Support Duo Win Rates

  nn - 0.580632565404139

  rf - 0.5734367155686952

  knn - 0.536620739666425

#Modeling on Champion, Bot/Support Duo, and Opposing Laner Win Rates

In [ ]:
champ_duo_bot_wr = pd.read_csv('match_champ_duo_bot_winrate.csv')
champ_duo_bot_wr

Match ID  Team A Won?     A Top  A Jungle     A Mid     A Bot  \
0       4848223288            1  0.521313  0.503173  0.503341  0.516784   
1       4848131748            1  0.479153  0.483957  0.505706  0.519829   
2       4848087528            1  0.521313  0.485958  0.493604  0.510774   
3       4847265470            0  0.549865  0.503162  0.506785  0.494525   
4       4848043757            1  0.493161  0.528064  0.514953  0.483096   
...            ...          ...       ...       ...       ...       ...   
128040  4814345455            1  0.487486  0.486515  0.520461  0.472028   
128041  4813002074            0  0.489415  0.470032  0.485156  0.522730   
128042  4819650090            0  0.521313  0.517958  0.503480  0.472028   
128043  4812981843            1  0.490364  0.505829  0.488907  0.527675   
128044  4814686156            1  0.501658  0.484444  0.485156  0.503527   

        A Support     B Top  B Jungle     B Mid     B Bot  B Support  \
0        0.472906  0.497863  0.505829  0.502770  0.522730   0.506754   
1        0.494923  0.502588  0.466134  0.520461  0.510774   0.506617   
2        0.512271  0.497370  0.508568  0.506992  0.507348   0.514714   
3        0.509598  0.490364  0.486515  0.503803  0.480917   0.506754   
4        0.517835  0.523136  0.495181  0.506785  0.482210   0.509598   
...           ...       ...       ...       ...       ...        ...   
128040   0.517835  0.490364  0.477379  0.464943  0.522730   0.514714   
128041   0.509598  0.510396  0.489342  0.471320  0.521180   0.478411   
128042   0.521975  0.502588  0.484444  0.496939  0.516784   0.494923   
128043   0.483742  0.518345  0.517958  0.505618  0.521180   0.483875   
128044   0.492683  0.523136  0.523916  0.500270  0.521180   0.491969   

           A Duo     B Duo  A Top MU  A Jungle MU  A Mid MU  A Bot MU  \
0       0.551724  0.554307  0.492611     0.408602  0.512658  0.502732   
1       0.536680  0.498225  0.493151     0.476190  0.436224  0.531835   
2       0.596154  0.550744  0.563536     0.544000  0.500000  0.462908   
3       0.510638  0.468321  0.400000     0.500000  0.468750  0.511128   
4       0.472785  0.524064  0.443038     0.535117  0.580952  0.513889   
...          ...       ...       ...          ...       ...       ...   
128040  0.433333  0.526749  0.465549     0.459184  0.500000  0.500000   
128041  0.493007  0.538462  0.520522     0.512195  0.519231  0.512456   
128042  0.250000  0.521595  0.507937     0.484848  0.428571  0.666667   
128043  0.495050  0.494253  0.435294     0.503226  0.416667  0.495356   
128044  0.428571  0.491379  0.563758     0.444444  0.608696  0.455285   

        A Support MU  
0           0.512821  
1           0.552239  
2           0.484507  
3           0.498871  
4           0.508224  
...              ...  
128040      0.484453  
128041      0.555556  
128042      0.537838  
128043      0.522901  
128044      0.480769  

[128045 rows x 19 columns]

In [ ]:
X = champ_duo_bot_wr.drop(['Match ID', 'Team A Won?'], axis=1)
y = champ_duo_bot_wr['Team A Won?']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=42)

X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size = 0.2, random_state=42)

In [ ]:
mlp = MLPClassifier(random_state=42).fit(X_train, y_train)
mlp.score(X_val, y_val)

0.649746192893401

In [ ]:
clf = RandomForestClassifier(max_depth = 5, random_state=42).fit(X_train, y_train)
clf.score(X_val, y_val)

0.6341830758074413

In [ ]:
knn = KNeighborsClassifier().fit(X_train, y_train)
knn.score(X_val, y_val)

0.6002119707703464

4. Modeling on Champion, Bot/Support Duo, and Opposing Laner Win Rates

  nn - 0.649746192893401

  rf - 0.6341830758074413

  knn - 0.6002119707703464

#Taking the best Dataframe and finding the best hyperparameters for each model

In [ ]:
X = champ_duo_bot_wr.drop(['Match ID', 'Team A Won?'], axis=1)
y = champ_duo_bot_wr['Team A Won?']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=42)

In [ ]:
param_dist = {
    'hidden_layer_sizes': [(50,), (100,), (50, 50), (100, 50)],
    'activation': ['logistic', 'tanh', 'relu'],
    'solver': ['lbfgs', 'sgd', 'adam'],
    'alpha': [1e-4, 1e-3, 1e-2, 1e-1],
    'learning_rate': ['constant', 'invscaling', 'adaptive'],
    'max_iter': randint(300, 500),
    'random_state': [42]
}

mlp = MLPClassifier()
mlp_cv = RandomizedSearchCV(mlp, param_distributions=param_dist, n_iter=10, cv=5, scoring='accuracy')
mlp_cv.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:541: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)
/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:541: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)
/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:541: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS R

RandomizedSearchCV(cv=5, estimator=MLPClassifier(),
                   param_distributions={'activation': ['logistic', 'tanh',
                                                       'relu'],
                                        'alpha': [0.0001, 0.001, 0.01, 0.1],
                                        'hidden_layer_sizes': [(50,), (100,),
                                                               (50, 50),
                                                               (100, 50)],
                                        'learning_rate': ['constant',
                                                          'invscaling',
                                                          'adaptive'],
                                        'max_iter': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7df70c23e140>,
                                        'random_state': [42],
                                        'solver': ['lbfgs', 'sgd', 'adam']},
                   scoring='accuracy')

In [ ]:
mlp_best_params = mlp_cv.best_params_
mlp_best_params

{'activation': 'relu',
 'alpha': 0.1,
 'hidden_layer_sizes': (100,),
 'learning_rate': 'constant',
 'max_iter': 312,
 'random_state': 42,
 'solver': 'lbfgs'}

In [ ]:
print(mlp_cv.best_score_)
mlp_test_accuracy = mlp_cv.score(X_test, y_test)

0.6493958903993728


In [ ]:
param_dist = {
    'n_estimators': randint(50, 200),
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': randint(2, 10),
    'min_samples_leaf': randint(1, 5),
    'bootstrap': [True, False],
    'random_state': [42]
}

rf = RandomForestClassifier()
rf_cv = RandomizedSearchCV(rf, param_distributions=param_dist, n_iter=10, cv=5, scoring='accuracy')
rf_cv.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomForestClassifiers and ExtraTreesClassifiers.
  warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:424: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features='sqrt'` or remove this parameter as it is also the default value for RandomFor

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(),
                   param_distributions={'bootstrap': [True, False],
                                        'max_depth': [None, 10, 20, 30],
                                        'max_features': ['auto', 'sqrt',
                                                         'log2'],
                                        'min_samples_leaf': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7df707402b00>,
                                        'min_samples_split': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7df707402290>,
                                        'n_estimators': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7df70bf07250>,
                                        'random_state': [42]},
                   scoring='accuracy')

In [ ]:
rf_best_params = rf_cv.best_params_
rf_best_params

{'bootstrap': True,
 'max_depth': 10,
 'max_features': 'log2',
 'min_samples_leaf': 1,
 'min_samples_split': 4,
 'n_estimators': 173,
 'random_state': 42}

In [ ]:
print(rf_cv.best_score_)
rf_test_accuracy = rf_cv.score(X_test, y_test)

0.6424786165117398


In [ ]:
param_dist = {
    'n_neighbors': [3, 5, 7, 10, 15],  # Example values for the number of neighbors
    'weights': ['uniform', 'distance'],  # Options for the weight function
    'metric': ['euclidean', 'manhattan', 'chebyshev']  # Distance metrics to consider
}

knn = KNeighborsClassifier()
knn_cv = RandomizedSearchCV(knn, param_distributions=param_dist, n_iter=10, cv=5, scoring='accuracy')
knn_cv.fit(X_train, y_train)

RandomizedSearchCV(cv=5, estimator=KNeighborsClassifier(),
                   param_distributions={'metric': ['euclidean', 'manhattan',
                                                   'chebyshev'],
                                        'n_neighbors': [3, 5, 7, 10, 15],
                                        'weights': ['uniform', 'distance']},
                   scoring='accuracy')

In [ ]:
knn_best_params = knn_cv.best_params_
knn_best_params

{'weights': 'uniform', 'n_neighbors': 15, 'metric': 'euclidean'}

In [ ]:
print(knn_cv.best_score_)
knn_test_accuracy = knn_cv.score(X_test, y_test)

0.6173756914073423


Neural Network - 0.6493958903993728

Random Forest - 0.6424786165117398

K-Nearest Neighbors - 0.6173756914073423